File: packages/skill-to-modal/pure_data_runtime.py
"""Trusted non-Turing-complete interpreter for reviewed pure-data workflows."""
from __future__ import annotations

import copy
import hashlib
import json
import re
from typing import Any


SPEC_VERSION = "omo.pure-data/v1"
STEP_ID_RE = re.compile(r"^[a-z][a-z0-9_]{0,63}$")
FIELD_RE = STEP_ID_RE
ASCII_WHITESPACE = " \t\r\n\f\v"
LIMIT_KEYS = {
    "max_steps",
    "max_input_bytes",
    "max_output_bytes",
    "max_list_items",
    "max_text_bytes",
}
LIMIT_RANGES = {
    "max_steps": (1, 16),
    "max_input_bytes": (1, 65_536),
    "max_output_bytes": (1, 65_536),
    "max_list_items": (1, 100),
    "max_text_bytes": (1, 1_000),
}
STEP_SHAPES = {
    "input.get": {"id", "op", "path"},
    "text_list.normalize_ascii": {
        "id", "op", "input", "trim_ascii_whitespace", "reject_empty", "reject_control_characters"
    },
    "text_list.unique": {"id", "op", "input", "comparison", "enabled_from"},
    "text_list.sort_ascii": {"id", "op", "input", "key", "tie_break"},
    "list.length": {"id", "op", "input"},
    "result.object": {"id", "op", "fields"},
}


def _canonical_bytes(value: Any) -> bytes:
    try:
        return json.dumps(
            value, ensure_ascii=True, allow_nan=False, separators=(",", ":"), sort_keys=True
        ).encode("ascii")
    except (TypeError, ValueError, UnicodeEncodeError) as exc:
        raise ValueError("NON_CANONICAL_VALUE") from exc


def canonical_pure_data_program(program: dict[str, Any]) -> str:
    return _canonical_bytes(program).decode("ascii")


def pure_data_program_digest(program: dict[str, Any]) -> str:
    return "sha256:" + hashlib.sha256(_canonical_bytes(program)).hexdigest()


def _require_closed_schema(schema: Any, name: str) -> dict[str, Any]:
    if not isinstance(schema, dict) or schema.get("type") != "object":
        raise ValueError(f"{name} must be an object schema")
    if schema.get("additionalProperties") is not False:
        raise ValueError(f"{name} must reject unknown fields")
    properties = schema.get("properties")
    required = schema.get("required")
    if not isinstance(properties, dict) or not isinstance(required, list):
        raise ValueError(f"{name} properties and required fields are mandatory")
    return schema


def _pointer_field(path: Any) -> str:
    if not isinstance(path, str) or not path.startswith("/") or path.count("/") != 1:
        raise ValueError("pure-data paths must name one top-level field")
    field = path[1:]
    if not FIELD_RE.fullmatch(field):
        raise ValueError("pure-data path field is invalid")
    return field


def _validate_limits(limits: Any) -> dict[str, int]:
    if not isinstance(limits, dict) or set(limits) != LIMIT_KEYS:
        raise ValueError("pure-data limits must use the closed v1 shape")
    result: dict[str, int] = {}
    for key, (minimum, maximum) in LIMIT_RANGES.items():
        value = limits.get(key)
        if not isinstance(value, int) or isinstance(value, bool) or not minimum <= value <= maximum:
            raise ValueError(f"pure-data limit {key} is invalid")
        result[key] = value
    return result


def validate_pure_data_program(
    program: dict[str, Any], input_schema: dict[str, Any], output_schema: dict[str, Any]
) -> dict[str, Any]:
    if not isinstance(program, dict) or set(program) != {"spec_version", "limits", "steps", "result"}:
        raise ValueError("pure-data program must use the closed v1 shape")
    if program.get("spec_version") != SPEC_VERSION:
        raise ValueError("pure-data spec version is unsupported")
    input_contract = _require_closed_schema(input_schema, "input_schema")
    output_contract = _require_closed_schema(output_schema, "output_schema")
    limits = _validate_limits(program.get("limits"))
    steps = program.get("steps")
    if not isinstance(steps, list) or not 1 <= len(steps) <= limits["max_steps"]:
        raise ValueError("pure-data steps exceed the reviewed bound")
    seen: dict[str, str] = {}
    result_fields: set[str] | None = None
    for index, step in enumerate(steps):
        if not isinstance(step, dict):
            raise ValueError("pure-data step must be an object")
        operation = step.get("op")
        if operation not in STEP_SHAPES:
            raise ValueError("pure-data operation is unsupported")
        if set(step) != STEP_SHAPES[operation]:
            raise ValueError("pure-data step shape is invalid")
        step_id = step.get("id")
        if not isinstance(step_id, str) or not STEP_ID_RE.fullmatch(step_id) or step_id in seen:
            raise ValueError("pure-data step id is invalid")
        if operation == "input.get":
            field = _pointer_field(step["path"])
            properties = input_contract["properties"]
            field_schema = properties.get(field)
            if field not in input_contract["required"] or not isinstance(field_schema, dict):
                raise ValueError("input.get must reference a required input field")
            seen[step_id] = "text_list" if field_schema.get("type") == "array" else "unknown"
        elif operation == "text_list.normalize_ascii":
            if seen.get(step["input"]) != "text_list":
                raise ValueError("normalize operation input reference is invalid")
            if any(step[key] is not True for key in (
                "trim_ascii_whitespace", "reject_empty", "reject_control_characters"
            )):
                raise ValueError("normalize operation policy must use safe v1 settings")
            seen[step_id] = "text_list"
        elif operation == "text_list.unique":
            if seen.get(step["input"]) != "text_list" or step.get("comparison") != "exact":
                raise ValueError("unique operation is invalid")
            condition = step.get("enabled_from")
            if not isinstance(condition, dict) or set(condition) != {"path", "default"}:
                raise ValueError("unique condition shape is invalid")
            field = _pointer_field(condition["path"])
            condition_schema = input_contract["properties"].get(field)
            if not isinstance(condition.get("default"), bool) or not isinstance(condition_schema, dict) or condition_schema.get("type") != "boolean":
                raise ValueError("unique condition must reference a boolean input")
            seen[step_id] = "text_list"
        elif operation == "text_list.sort_ascii":
            if (
                seen.get(step["input"]) != "text_list"
                or step.get("key") != "ascii_case_insensitive"
                or step.get("tie_break") != "ascii_bytes"
            ):
                raise ValueError("sort operation is invalid")
            seen[step_id] = "text_list"
        elif operation == "list.length":
            if seen.get(step["input"]) != "text_list":
                raise ValueError("length operation input reference is invalid")
            seen[step_id] = "integer"
        else:
            if index != len(steps) - 1:
                raise ValueError("result.object must be the final step")
            fields = step.get("fields")
            output_properties = output_contract["properties"]
            if not isinstance(fields, dict) or set(fields) != set(output_properties) or len(fields) > 32:
                raise ValueError("result.object fields must match output schema")
            for field, descriptor in fields.items():
                if not FIELD_RE.fullmatch(field) or not isinstance(descriptor, dict) or len(descriptor) != 1:
                    raise ValueError("result.object field shape is invalid")
                if "ref" in descriptor:
                    reference = descriptor["ref"]
                    if not isinstance(reference, str) or reference not in seen:
                        raise ValueError("result.object reference is invalid")
                elif "const" in descriptor:
                    value = descriptor["const"]
                    valid_constant = (
                        value is None
                        or isinstance(value, bool)
                        or (
                            isinstance(value, int)
                            and not isinstance(value, bool)
                            and abs(value) <= 9_007_199_254_740_991
                        )
                        or (
                            isinstance(value, str)
                            and all(32 <= ord(char) <= 126 for char in value)
                            and len(value.encode("ascii")) <= 200
                        )
                    )
                    if not valid_constant:
                        raise ValueError("result.object constant is invalid")
                else:
                    raise ValueError("result.object field source is invalid")
            seen[step_id] = "object"
            result_fields = set(fields)
    result_id = program.get("result")
    if not isinstance(result_id, str) or seen.get(result_id) != "object" or result_id != steps[-1]["id"]:
        raise ValueError("pure-data result reference is invalid")
    if result_fields != set(output_contract["required"]):
        raise ValueError("pure-data output schema must require every result field")
    return copy.deepcopy(program)


def _ascii_text_list(value: Any, limits: dict[str, int]) -> list[str]:
    if not isinstance(value, list) or not 0 <= len(value) <= limits["max_list_items"]:
        raise ValueError("INVALID_INPUT")
    result: list[str] = []
    for item in value:
        if not isinstance(item, str):
            raise ValueError("INVALID_INPUT")
        try:
            encoded = item.encode("ascii")
        except UnicodeEncodeError as exc:
            raise ValueError("INVALID_VALUE") from exc
        if len(encoded) > limits["max_text_bytes"]:
            raise ValueError("INVALID_VALUE")
        result.append(item)
    return result


def execute_pure_data_program(program: dict[str, Any], payload: dict[str, Any]) -> dict[str, Any]:
    limits = _validate_limits(program.get("limits"))
    if not isinstance(payload, dict):
        raise ValueError("INVALID_INPUT")
    if len(_canonical_bytes(payload)) > limits["max_input_bytes"]:
        raise ValueError("INPUT_LIMIT_EXCEEDED")
    values: dict[str, Any] = {}
    for step in program["steps"]:
        operation = step["op"]
        if operation == "input.get":
            field = _pointer_field(step["path"])
            values[step["id"]] = _ascii_text_list(payload.get(field), limits)
        elif operation == "text_list.normalize_ascii":
            normalized = [item.strip(ASCII_WHITESPACE) for item in values[step["input"]]]
            if any(not item or any(ord(char) < 32 or ord(char) == 127 for char in item) for item in normalized):
                raise ValueError("INVALID_VALUE")
            values[step["id"]] = normalized
        elif operation == "text_list.unique":
            condition = step["enabled_from"]
            enabled = payload.get(_pointer_field(condition["path"]), condition["default"])
            if not isinstance(enabled, bool):
                raise ValueError("INVALID_INPUT")
            source = values[step["input"]]
            values[step["id"]] = list(dict.fromkeys(source)) if enabled else list(source)
        elif operation == "text_list.sort_ascii":
            source = values[step["input"]]
            values[step["id"]] = sorted(source, key=lambda item: (item.lower().encode("ascii"), item.encode("ascii")))
        elif operation == "list.length":
            values[step["id"]] = len(values[step["input"]])
        elif operation == "result.object":
            result: dict[str, Any] = {}
            for field, descriptor in step["fields"].items():
                result[field] = descriptor["const"] if "const" in descriptor else copy.deepcopy(values[descriptor["ref"]])
            values[step["id"]] = result
        else:
            raise ValueError("PROGRAM_OPERATION_INVALID")
    output = values[program["result"]]
    if not isinstance(output, dict) or len(_canonical_bytes(output)) > limits["max_output_bytes"]:
        raise ValueError("OUTPUT_LIMIT_EXCEEDED")
    return copy.deepcopy(output)

File: site/deploy/pure-data-runtime.mjs
const PROGRAM_SPEC_VERSION = 'omo.pure-data/v1';
const HARD_LIMITS = Object.freeze({
  max_steps: 16,
  max_input_bytes: 64 * 1024,
  max_output_bytes: 64 * 1024,
  max_list_items: 100,
  max_text_bytes: 1000,
});
const LIMIT_KEYS = Object.freeze(Object.keys(HARD_LIMITS));
const STEP_ID_RE = /^[A-Za-z][A-Za-z0-9_]{0,63}$/;
const FIELD_NAME_RE = /^[A-Za-z][A-Za-z0-9_]{0,63}$/;
const encoder = new TextEncoder();

export class PureDataRuntimeError extends Error {
  constructor(code, message) {
    super(`${code}: ${message}`);
    this.name = 'PureDataRuntimeError';
    this.code = code;
  }
}

function fail(code, message) {
  throw new PureDataRuntimeError(code, message);
}

function isRecord(value) {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}

function hasExactKeys(value, allowed) {
  if (!isRecord(value)) return false;
  const actual = Object.keys(value).sort();
  const expected = [...allowed].sort();
  return actual.length === expected.length && actual.every((key, index) => key === expected[index]);
}

function hasAllowedKeys(value, required, optional = []) {
  if (!isRecord(value)) return false;
  const allowed = new Set([...required, ...optional]);
  return required.every((key) => Object.prototype.hasOwnProperty.call(value, key))
    && Object.keys(value).every((key) => allowed.has(key));
}

function assertPositiveBoundedInteger(value, key) {
  if (!Number.isInteger(value) || value < 1 || value > HARD_LIMITS[key]) {
    fail('INVALID_PROGRAM', `invalid or unsafe ${key}`);
  }
}

function assertJsonValue(value, label) {
  if (value === null || typeof value === 'string' || typeof value === 'boolean') return;
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) fail('INVALID_PROGRAM', `${label} is not finite JSON data`);
    return;
  }
  if (Array.isArray(value)) {
    for (const child of value) assertJsonValue(child, label);
    return;
  }
  if (isRecord(value)) {
    for (const [key, child] of Object.entries(value)) {
      if (key === '__proto__' || key === 'prototype' || key === 'constructor') {
        fail('INVALID_PROGRAM', `${label} contains an unsafe key`);
      }
      assertJsonValue(child, label);
    }
    return;
  }
  fail('INVALID_PROGRAM', `${label} is not JSON data`);
}

function assertCanonicalConstant(value, label) {
  if (value === null || typeof value === 'boolean') return;
  if (typeof value === 'number' && Number.isSafeInteger(value) && !Object.is(value, -0)) return;
  if (typeof value === 'string' && /^[\x20-\x7E]{0,200}$/.test(value)) return;
  fail('INVALID_PROGRAM', `${label} constant is outside the canonical subset`);
}

function assertPointer(path, label) {
  if (typeof path !== 'string' || !/^\/[A-Za-z][A-Za-z0-9_]{0,63}$/.test(path)) {
    fail('INVALID_PROGRAM', `${label} must name one top-level input field`);
  }
}

function assertReference(id, knownIds, label) {
  if (typeof id !== 'string' || !knownIds.has(id)) {
    fail('INVALID_PROGRAM', `${label} must reference an earlier step`);
  }
}

function validateStep(step, knownIds) {
  if (!isRecord(step) || !STEP_ID_RE.test(String(step.id || '')) || knownIds.has(step.id)) {
    fail('INVALID_PROGRAM', 'step id shape is invalid or duplicated');
  }
  switch (step.op) {
    case 'input.get':
      if (!hasExactKeys(step, ['id', 'op', 'path'])) fail('INVALID_PROGRAM', 'input.get step shape is invalid');
      assertPointer(step.path, 'input.get path');
      break;
    case 'text_list.normalize_ascii':
      if (!hasExactKeys(step, [
        'id', 'op', 'input', 'trim_ascii_whitespace', 'reject_empty', 'reject_control_characters',
      ])) fail('INVALID_PROGRAM', 'text_list.normalize_ascii step shape is invalid');
      assertReference(step.input, knownIds, 'text_list.normalize_ascii input');
      for (const key of ['trim_ascii_whitespace', 'reject_empty', 'reject_control_characters']) {
        if (typeof step[key] !== 'boolean') fail('INVALID_PROGRAM', `${key} must be boolean`);
      }
      break;
    case 'text_list.unique':
      if (!hasAllowedKeys(step, ['id', 'op', 'input', 'comparison'], ['enabled_from'])) {
        fail('INVALID_PROGRAM', 'text_list.unique step shape is invalid');
      }
      assertReference(step.input, knownIds, 'text_list.unique input');
      if (step.comparison !== 'exact') fail('INVALID_PROGRAM', 'text_list.unique comparison is unsupported');
      if (Object.prototype.hasOwnProperty.call(step, 'enabled_from')) {
        if (!hasExactKeys(step.enabled_from, ['path', 'default']) || typeof step.enabled_from.default !== 'boolean') {
          fail('INVALID_PROGRAM', 'text_list.unique enabled_from shape is invalid');
        }
        assertPointer(step.enabled_from.path, 'enabled_from path');
      }
      break;
    case 'text_list.sort_ascii':
      if (!hasExactKeys(step, ['id', 'op', 'input', 'key', 'tie_break'])) {
        fail('INVALID_PROGRAM', 'text_list.sort_ascii step shape is invalid');
      }
      assertReference(step.input, knownIds, 'text_list.sort_ascii input');
      if (step.key !== 'ascii_case_insensitive' || step.tie_break !== 'ascii_bytes') {
        fail('INVALID_PROGRAM', 'text_list.sort_ascii ordering is unsupported');
      }
      break;
    case 'list.length':
      if (!hasExactKeys(step, ['id', 'op', 'input'])) fail('INVALID_PROGRAM', 'list.length step shape is invalid');
      assertReference(step.input, knownIds, 'list.length input');
      break;
    case 'result.object':
      if (!hasExactKeys(step, ['id', 'op', 'fields']) || !isRecord(step.fields)) {
        fail('INVALID_PROGRAM', 'result.object step shape is invalid');
      }
      for (const [field, source] of Object.entries(step.fields)) {
        if (!FIELD_NAME_RE.test(field)) fail('INVALID_PROGRAM', 'result.object field name is invalid');
        if (hasExactKeys(source, ['ref'])) assertReference(source.ref, knownIds, `result.object ${field}`);
        else if (hasExactKeys(source, ['const'])) assertCanonicalConstant(source.const, `result.object ${field}`);
        else fail('INVALID_PROGRAM', `result.object ${field} source shape is invalid`);
      }
      break;
    default:
      fail('INVALID_PROGRAM', `operation ${String(step.op || '')} is unsupported`);
  }
  knownIds.add(step.id);
}

export function validatePureDataProgram(program) {
  if (!hasExactKeys(program, ['spec_version', 'limits', 'steps', 'result'])) {
    fail('INVALID_PROGRAM', 'program shape is invalid');
  }
  if (program.spec_version !== PROGRAM_SPEC_VERSION) fail('INVALID_PROGRAM', 'program spec version is unsupported');
  if (!hasExactKeys(program.limits, LIMIT_KEYS)) fail('INVALID_PROGRAM', 'limits shape is invalid');
  for (const key of LIMIT_KEYS) assertPositiveBoundedInteger(program.limits[key], key);
  if (!Array.isArray(program.steps) || program.steps.length < 1 || program.steps.length > program.limits.max_steps) {
    fail('INVALID_PROGRAM', 'step count exceeds reviewed limits');
  }
  const knownIds = new Set();
  for (const step of program.steps) validateStep(step, knownIds);
  if (typeof program.result !== 'string' || !knownIds.has(program.result)) {
    fail('INVALID_PROGRAM', 'program result must reference a step');
  }
  const resultStep = program.steps.find((step) => step.id === program.result);
  if (!resultStep || resultStep.op !== 'result.object') fail('INVALID_PROGRAM', 'program result must be result.object');
  return program;
}

function canonicalJson(value) {
  if (value === null || typeof value !== 'object') return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(',')}]`;
  return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${canonicalJson(value[key])}`).join(',')}}`;
}

export async function pureDataProgramDigest(program) {
  validatePureDataProgram(program);
  const digest = await globalThis.crypto.subtle.digest('SHA-256', encoder.encode(canonicalJson(program)));
  const hex = Array.from(new Uint8Array(digest), (byte) => byte.toString(16).padStart(2, '0')).join('');
  return `sha256:${hex}`;
}

function pointerGet(root, pointer, fallbackMarker) {
  let current = root;
  for (const encodedPart of pointer.slice(1).split('/')) {
    const part = encodedPart.replace(/~1/g, '/').replace(/~0/g, '~');
    if (!isRecord(current) && !Array.isArray(current)) return fallbackMarker;
    if (!Object.prototype.hasOwnProperty.call(current, part)) return fallbackMarker;
    current = current[part];
  }
  return current;
}

function jsonBytes(value) {
  let serialized;
  try { serialized = JSON.stringify(value); } catch { fail('INVALID_VALUE', 'value is not serializable JSON'); }
  if (serialized === undefined) fail('INVALID_VALUE', 'value is not serializable JSON');
  return encoder.encode(serialized).length;
}

function assertTextList(value, limits, label) {
  if (!Array.isArray(value)) fail('INVALID_VALUE', `${label} must be a list`);
  if (value.length > limits.max_list_items) fail('LIST_LIMIT_EXCEEDED', `${label} has too many items`);
  for (const item of value) {
    if (typeof item !== 'string') fail('INVALID_VALUE', `${label} items must be text`);
    if (encoder.encode(item).length > limits.max_text_bytes) fail('TEXT_LIMIT_EXCEEDED', `${label} item is too large`);
  }
}

function asciiCompare(left, right) {
  const leftFolded = left.replace(/[A-Z]/g, (character) => character.toLowerCase());
  const rightFolded = right.replace(/[A-Z]/g, (character) => character.toLowerCase());
  if (leftFolded < rightFolded) return -1;
  if (leftFolded > rightFolded) return 1;
  if (left < right) return -1;
  if (left > right) return 1;
  return 0;
}

function cloneJson(value) {
  return JSON.parse(JSON.stringify(value));
}

export function executePureDataProgram(program, input) {
  validatePureDataProgram(program);
  if (!isRecord(input)) fail('INVALID_VALUE', 'input must be an object');
  const limits = program.limits;
  if (jsonBytes(input) > limits.max_input_bytes) fail('INPUT_LIMIT_EXCEEDED', 'input exceeds reviewed byte limit');
  const values = new Map();
  const missing = Symbol('missing');

  for (const step of program.steps) {
    let value;
    switch (step.op) {
      case 'input.get':
        value = pointerGet(input, step.path, missing);
        if (value === missing) fail('INVALID_VALUE', `input path ${step.path} is missing`);
        value = cloneJson(value);
        break;
      case 'text_list.normalize_ascii': {
        const source = values.get(step.input);
        assertTextList(source, limits, step.input);
        value = source.map((item) => {
          let normalized = step.trim_ascii_whitespace ? item.replace(/^[\x09-\x0D\x20]+|[\x09-\x0D\x20]+$/g, '') : item;
          if (/[^\x00-\x7F]/.test(normalized)) fail('INVALID_VALUE', 'text_list.normalize_ascii requires ASCII text');
          if (step.reject_control_characters && /[\x00-\x1F\x7F]/.test(normalized)) {
            fail('INVALID_VALUE', 'text_list.normalize_ascii rejected a control character');
          }
          if (step.reject_empty && normalized.length === 0) fail('INVALID_VALUE', 'text_list.normalize_ascii rejected empty text');
          if (encoder.encode(normalized).length > limits.max_text_bytes) fail('TEXT_LIMIT_EXCEEDED', 'normalized text is too large');
          return normalized;
        });
        break;
      }
      case 'text_list.unique': {
        const source = values.get(step.input);
        assertTextList(source, limits, step.input);
        const enabled = step.enabled_from
          ? pointerGet(input, step.enabled_from.path, step.enabled_from.default)
          : true;
        if (typeof enabled !== 'boolean') fail('INVALID_VALUE', 'enabled_from value must be boolean');
        value = enabled ? [...new Set(source)] : [...source];
        break;
      }
      case 'text_list.sort_ascii': {
        const source = values.get(step.input);
        assertTextList(source, limits, step.input);
        for (const item of source) if (/[^\x00-\x7F]/.test(item)) fail('INVALID_VALUE', 'text_list.sort_ascii requires ASCII text');
        value = [...source].sort(asciiCompare);
        break;
      }
      case 'list.length': {
        const source = values.get(step.input);
        if (!Array.isArray(source)) fail('INVALID_VALUE', `${step.input} must be a list`);
        if (source.length > limits.max_list_items) fail('LIST_LIMIT_EXCEEDED', `${step.input} has too many items`);
        value = source.length;
        break;
      }
      case 'result.object':
        value = {};
        for (const [field, source] of Object.entries(step.fields)) {
          value[field] = Object.prototype.hasOwnProperty.call(source, 'ref')
            ? cloneJson(values.get(source.ref))
            : cloneJson(source.const);
        }
        break;
      default:
        fail('INVALID_PROGRAM', 'unreachable unsupported operation');
    }
    values.set(step.id, value);
  }

  const result = values.get(program.result);
  if (!isRecord(result)) fail('INVALID_VALUE', 'program result is not an object');
  if (jsonBytes(result) > limits.max_output_bytes) fail('OUTPUT_LIMIT_EXCEEDED', 'output exceeds reviewed byte limit');
  return result;
}
